internal/pkg/skuba/addons/addons.go

```diff
@@ -0,0 +1,288 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package addons
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"text/template"
+
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/klog"
+
+	"github.com/SUSE/skuba/internal/pkg/skuba/kubernetes"
+	"github.com/SUSE/skuba/internal/pkg/skuba/skuba"
+	skubaconstants "github.com/SUSE/skuba/pkg/skuba"
+	"github.com/pkg/errors"
+)
+
+type addonPriority uint
+
+const (
+	highPriority   addonPriority = iota
+	normalPriority addonPriority = iota
+)
+
+var Addons = map[kubernetes.Addon]Addon{}
+
+type Addon struct {
+	addon         kubernetes.Addon
+	templater     addonTemplater
+	callbacks     addonCallbacks
+	addonPriority addonPriority
+}
+
+type addonCallbacks interface {
+	BeforeApply(AddonConfiguration, *skuba.SkubaConfiguration) error
+	AfterApply(AddonConfiguration, *skuba.SkubaConfiguration) error
+}
+
+type AddonConfiguration struct {
+	ClusterVersion *version.Version
+	ControlPlane   string
+	ClusterName    string
+}
+
+type addonTemplater func(AddonConfiguration) string
+
+type ApplyBehavior uint
+
+const (
+	// This is the default behavior for all operations except for Bootstrap,
+	// the addon is always re-rendered prior to being applied. In an addons
+	// upgrade operation for example, we always want to re-render the latest
+	// contents and never reuse local file contents in case the upgrade was
+	// executed inside a cluster definition folder
+	AlwaysRender ApplyBehavior = iota
+	// This is the desired behavior for Bootstrap, when the user can tweak
+	// the addon configurations after `skuba cluster init`, so
+	// `skuba node bootstrap` will apply the modified settings instead of
+	// re-rendering them forcefully
+	SkipRenderIfConfigFilePresent ApplyBehavior = iota
+)
+
+type renderContext struct {
+	addon  Addon
+	config AddonConfiguration
+}
+
+func (renderContext renderContext) AnnotatedVersion() string {
+	return fmt.Sprintf("addon.caasp.suse.com/manifest-version: \"%s\"", renderContext.ManifestVersion())
+}
+
+func (renderContext renderContext) ManifestVersion() string {
+	addonVersion := kubernetes.AddonVersionForClusterVersion(renderContext.addon.addon, renderContext.config.ClusterVersion)
+	if addonVersion == nil {
+		return ""
+	}
+	return fmt.Sprintf("%s-%d", addonVersion.Version, addonVersion.ManifestVersion)
+}
+
+func registerAddon(addon kubernetes.Addon, addonTemplater addonTemplater, callbacks addonCallbacks, addonPriority addonPriority) {
+	Addons[addon] = Addon{
+		addon:         addon,
+		templater:     addonTemplater,
+		callbacks:     callbacks,
+		addonPriority: addonPriority,
+	}
+}
+
+func addonsByPriority() []Addon {
+	sortedAddons := make([]Addon, len(Addons))
+	i := 0
+	for _, addon := range Addons {
+		sortedAddons[i] = addon
+		i++
+	}
```

Comment:
Nitpicking: just:

```go
	for i, addon := range Addons {
		sortedAddons[i] = addon
	}
```

Since both slices have the same length, there is no need for that `i` variable outside of the for loop.

Reply:
But `Addons` is a map though, the first argument would be the `kubernetes.Addon` instead of the iteration.

Reply:
Ah, missed that part. Sorry...